src/split-by.mjs

```diff
@@ -0,0 +1,93 @@
+import { iterableCurry } from './internal/iterable'
+import { Exchange } from './internal/queues'
+
+function splitBy (getKey = (k) => k, iterable) {
```

Comment:
Couldn't you replace this whole thing with `map(([,group]) => group, groupBy)`

But also what happened to the splitBy we discussed which had a predicate callback that identified separator items which would not be present in the final output?

Reply:
groupBy and splitBy are similar, but also different.

In splitBy every single iterator emitted is responsible to consume the original iterator.

In groupBy, there is a "main loop" that looks ahead to search for different keys.

This also changes the output of asyncGroupBy. In fact this returns an asyncIterable of asyncIterables while asyncSplitBy returns an iterable of asyncIterables

Reply:
I'm really still not understanding. How can it synchronously know how many groups it is going to create out of an asynchronous iterable? Or, looking at the test that exist, it requires the caller to know how many groups they expect? That seems like an API flaw to me more than a feature. And conceptually it is still grouping.

I could get the result easily enough (for sync things) then like this:

const [[, As], [, Bs] = groupBy("AABB")

Reply:
> How can it synchronously know how many groups it is going to create out of an asynchronous iterable?

It doesn't. Just like fork.
If you are interested in working with all groups you will have to consume one at the time until it returns an empty group.

Reply:
Why would this be a better API than the one we had discussed though?

Reply:
I'm working on an implementation of splitBy that matches String.prototype.split functionality. I think it is well worthwhile not to force the user to handle the termination condition themselves, which would make it virtually impossible to chain the method.

Reply:
> Why would this be a better API than the one we had discussed though?

I designed splitBy to be a low level implementation for any case where you slice an iterable. You can build any other case on top of it (as I did for splitAt).

Reply:
> make it virtually impossible to chain the method

it seems to be a bit an overstatement to me. I am sure there are plenty of cases where this would do the trick.

Reply:
My biggest complaint is that it doesn't seem to have anything to do with splitting. It's a low level grouping function. To split something, you must be able to choose the point at which to split it. That is either a size 1 match (item => splitOnItem) or a size 0 match (a, b) => splitBetweenAandB. To group you define a characteristic for the group, and items sharing the characteristic are part of the group, which is what this function is about.

That would mean that we still need a splitting method, which I have written [here](https://gist.github.com/conartist6/e3e7e19fde884b920fca7359163640ec#file-split-by-mjs).

Reply:
If you want, I'll make the async version and turn that gist into a PR on this PR

Reply:
> My biggest complaint is that it doesn't seem to have anything to do with splitting

We can still come up with a different name.

> If you want, I'll make the async version and turn that gist into a PR on this PR

No please wait for it.
As I said I'd like to use this as a low level abstraction for these family of functions. I don't want to maintain 3/4 functions that are almost the same.

Reply:
Yeah it would be great if these functions could share some more code. I can work on trying to get my implementations to share some, but to do that I'd first need to know that you want a split function with that API.

Reply:
I just don't want to work backwards from internal code sharing to the external API

Reply:
The way you implemented splitBy is very similar to groupBy. The main loop consumes the original iterator (this is why you are able to know when to stop producing iterators).

This way, however, you are unable to reimplement asyncSplitAt, because you can't do this:
```js
const [iter1, iter2, iter3] = asyncSplitBy(iterable) // asyncIterable can't be spread
```

That is exactly the reason why I think both groupBy and splitBy (the way I implemented, replace with any name you prefer ) covers 2 different and valuable use cases.

Reply:
Couldn't splitAt be implemented with plain old `groupBy` though?
```js
function splitAt(index, iterable) {
  const [[, first]. [, second]] = groupBy((_, i) => i < index ? 0 : 1, iterable)
  return [first, second]
}
```

Reply:
Ah but you're saying specifically that that doesn't work for async, I see.

Reply:
yep

Reply:
But that's still pretty easy to get around.
```js
function asyncSplitAt(index, iterable) {
  const groups = groupBy((_, i) => i < index ? 0 : 1, iterable)
  let firstGroup = groups.next();
  let secondGroup = groups.next();
  async function * first() {
    yield * await firstGroup
  }
  async function * second() {
    yield * await secondGroup()
  }
  return [first(), second()]
}
```

Reply:
"await" in a sync function ?

Reply:
Yeah my bad see revised version.

Reply:
Let me think. If we can suppress splitBy, in favour of groupBy, I am more than happy to do so

Reply:
I guess in theory we could implement splitBy with groupBy as well I suppose?
```js
splitBy(predicate, iterable) {
  const grouped = groupBy((item, i) => !!predicate(item, i), iterable)
  return filter((_, i) => isEven(i), grouped)
}
```

Reply:
That won't work though if you have two separators next to each other, in which case you should get an empty group. Anyway I still think this method is premature, and it's the last thing stopping me from saying that you should go ahead and merge this.